### PR TITLE
deleting breaking remove of helmx.1.rendered dir

### PR DESCRIPTION
**Pull request by tomassatka**
Before the change when applying transformers user got following error from kustomize binary that kustomize failed and the reason was that in kustomization.yaml file were referred files inside helmx.1 directory however the directory was not existing at that point of time
```bash
tomas@tomas-ThinkPad-W550s:~/Develop/test$ helmfile template 
in ./helmfile.yaml: [exit status 1

COMMAND:
  kustomize build /tmp/chartify043502074/test --output /tmp/chartify043502074/test/all.patched.yaml

OUTPUT:
  Error: accumulating resources: accumulation err='accumulating resources from 'helmx.1.rendered/test/templates/serviceaccount.yaml': evalsymlink failure on '/tmp/chartify043502074/test/helmx.1.rendered/test/templates/serviceaccount.yaml' : lstat /tmp/chartify043502074/test/helmx.1.rendered: no such file or directory': evalsymlink failure on '/tmp/chartify043502074/test/helmx.1.rendered/test/templates/serviceaccount.yaml' : lstat /tmp/chartify043502074/test/helmx.1.rendered: no such file or directory]
```

solution was to remove the os.RemoveAll(helmOutputDir) poiting to helmx.1 dir 

After removing and local testing the transformers works again. 

Please consider to apply and include into helmfile as updated version of go dependency

### CODE REVIEW

**Comment by mumoshu** [comment]
@tomassatka Hey! Thanks for your efforts ☺️ Would you mind sharing your helmfile.yaml for reproduction? I thought it worked before so this might be either a regression or some edge-case that we didn't encounter before. A example helmfile.yaml helps isolatating the cause, so that I can better review this. Thanks!

**Comment by tomassatka** [comment]
hi @mumoshu 

**helm** version : version.BuildInfo{Version:"v3.7.1", GitCommit:"1d11fcb5d3f3bf00dbe6fe31b8412839a96b3dc4", GitTreeState:"clean", GoVersion:"go1.16.9"}
**helmfile** version : helmfile version v0.142.0
**kustomize** version : {Version:kustomize/v4.4.0 GitCommit:63ec6bdb3d737a7c66901828c5743656c49b60e1 BuildDate:2021-09-27T16:24:12Z GoOs:linux GoArch:amd64}

i took the most simple test case i could think of

created simple helm chart called test by `helm create test`

then added simple helmfile.yaml
```yaml
releases:
  - name: test
    chart: ./ 
    version: 0.1.0
    transformers: 
    - ./transformer.yaml 
```
and simple transformer.yaml 
```yaml
apiVersion: builtin
kind: PatchTransformer
metadata:
  name: test-transformer
patch: |-
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: test
    labels:
      test-transformer: did-my-job
target:
  kind: Deployment
  name: test

```

feel free to try out on your own

i also fixed few parts for tests like chart repo url and minor parts around. I dont know how to fix the tests since i dont know the expected behavior.. but at least tests now run during git event..

**Comment by mumoshu** [comment]
@tomassatka Thanks. I was able to reproduce this. Let me review this real quick to see what's happening and how your fix works.

**Comment by mumoshu** [comment]
@tomassatka Hey. It turned out you found an edge-case!

So it seems to break only when you had `chart: ./`. It just work if you had placed your chart under somewhere else.
For example, the below works if you created the chart under `test/test` and helmfile config is at `test/helmfile.yaml`

```
$ tree .
.
├── charts
├── Chart.yaml
├── helmfile.yaml
├── templates
│   ├── deployment.yaml
│   ├── _helpers.tpl
│   ├── hpa.yaml
│   ├── ingress.yaml
│   ├── NOTES.txt
│   ├── serviceaccount.yaml
│   ├── service.yaml
│   └── tests
│       └── test-connection.yaml
├── test
│   ├── charts
│   ├── Chart.yaml
│   ├── templates
│   │   ├── deployment.yaml
│   │   ├── _helpers.tpl
│   │   ├── hpa.yaml
│   │   ├── ingress.yaml
│   │   ├── NOTES.txt
│   │   ├── serviceaccount.yaml
│   │   ├── service.yaml
│   │   └── tests
│   │       └── test-connection.yaml
│   └── values.yaml
├── transformer.yaml
└── values.yaml
```

```
releases:
  - name: test
    chart: ./test
    version: 0.1.0
    transformers: 
    - ./transformer.yaml 
```

File: replace.go [hunk 1]
```diff
@@ -149,10 +149,6 @@ func (r *Runner) ReplaceWithRendered(name, chartName, chartPath string, o Replac
 		return nil, fmt.Errorf("invalid state: no files rendered")
 	}
 
-	if err := os.RemoveAll(helmOutputDir); err != nil {
-		return nil, fmt.Errorf("cleaning up unnecessary files after replace: %v", err)
-	}
```
Comment (mumoshu) [inline replace.go:154]: We still need this block for cleanup when your chart is located at a place other than `./`.

https://github.com/variantdev/chartify/pull/13#issuecomment-1013647811

See L139-L141- what it does there is to rename anything under `$TEMP_DIR/helmx1.rendered/$SUB_DIR` to `$TEMP_DIR/$SUB_DIR`, so that we can effectively "replace" the chart contained in `$TEMP_DIR` with the rendering result of the chart. We only use `$TEMP_DIR/helmx1.rendered` for a temporary location of the intermediate chart output, which can be removed after L147.

But `chartName` of `.` breaks that's behavior. The important part is L70-L76:

```
	helmOutputDir := filepath.Join(chartPath, "helmx.1.rendered")
	if err := os.Mkdir(helmOutputDir, 0755); err != nil {
		return nil, err
	}
	// This directory contains templates/ and charts/SUBCHART/templates
	chartOutputDir := filepath.Join(helmOutputDir, chartName)
```

So this function works correct only when `helmOutputDir != chartOutputDir`, because we move everything under `helmOutputDir` to `chartOutputDir` and remove `helmOutputDir`.

But `chartName` of `.` makes it `helmOutputDir == chartOutputDir`, which ends up removing the whole output dir which must exist for later process.

Comment (mumoshu) [reply replace.go:154]: I believe the correct fix would be to modify L76 instead.

More concretely, in the below code:

```
chartOutputDir := filepath.Join(helmOutputDir, chartName)
```

`chartOuputDir` shouldn't result in `helmOutputDir` if `chartName` is `.`.
Probably we'd better get the chartName by parsing `Chart.yaml` instead? Probably that's what `helm template --output-dir` appends to the output dir path.

Comment (tomassatka) [reply replace.go:154]: Hi @mumoshu 
Thats an interesting edge case but honestly i was not looking for that. I am facing same symptoms on windows when trying to use transformers and by debugging and using simple test case on linux i thought i found the rootcause of the issue. I would have to continue to debug also on windows to understand what could be the problem there. Would you mind to keep this topic open?

thx 
tom

Comment (icy) [reply replace.go:154]: > Probably we'd better get the chartName by parsing `Chart.yaml` instead? Probably that's what `helm template --output-dir` appends to the output dir path.

If chartName is not from Chart.yaml, I think that explains my problem in https://github.com/variantdev/chartify/pull/13#issuecomment-1039090849

Comment (icy) [reply replace.go:154]: See also https://github.com/helm/helm-www/blob/5ce1672465102890a16a3f9ef6549444084f6df1/content/en/docs/topics/charts.md#the-chart-file-structure

**Comment by mumoshu** [comment]
@icy Would you mind sharing the relevant part of your helmfile? This bug affects things when your chart path's basename is `.`. So it would work with `chart: ./foobar` but not work for `chart: ./foobar/.` or `chart: ./` or `chart:. `.

**Comment by icy** [comment]
> @icy Would you mind sharing the relevant part of your helmfile? This bug affects things when your chart path's basename is `.`. So it would work with `chart: ./foobar` but not work for `chart: ./foobar/.` or `chart: ./` or `chart:. `.

Hi @mumoshu, you can fetch the test code from this repo.

https://github.com/icy/helmfile-bug/tree/main/two

I used helmfile v0.143.0. In the code [1], if you rename the the chartName to `test` the issue is gone.

Please let me know if you need any further information. Thanks.

[1] https://github.com/icy/helmfile-bug/blob/main/two/charts/test/Chart.yaml#L1

File: replace.go (hunk 1 above)
Comment (mumoshu) [reply replace.go:154]: Thanks for sharing.
An another important point for this is that there can be cases that `Chart.yaml` is missing, for example when you're turning raw k8s manifests or kustomize config into a chart using chartify.
Perhaps a better approach would be to use use the last non empty basename of the path(so if chart is `foo/bar/../.` it should use `bar` as the chart name)

**Comment by mumoshu** [comment]
@tomassatka @icy I finally got some time to work on what we've discussed in the thread and the result is #27. I was able to reproduce the issue using the new integration tests and confirmed the issue was fixed by the use of `filepath.Clean` before taking the basename of the directory.
I'd appreciate it if you could give it a shot by building helmfile yourself replacing the chartify dependency using `replace github.com/variantdev/chartify => ../chartify` in helmfile's go.mod, or once I cut a new patch version of Helmfile(Not sure when though!

**Comment by icy** [comment]
@mumoshu Thanks for your work. I have given a test with my test https://github.com/icy/helmfile-bug/tree/main/two and the problem still remains, maybe I missed sth, or my problem is another one?

```
$ pwd
/home/gfg/go/src/github.com/roboll/helmfile

$ git clog -1
7423913

$ tail -1 go.mod
replace github.com/variantdev/chartify => ../chartify

$ (cd ../chartify && git clog -1 )
7b5aa80

$ go build

$ cd ~/projects/helmfile-bug/two

$ /home/gfg/go/src/github.com/roboll/helmfile/helmfile template 
in ./helmfile.yaml: [exit status 1

COMMAND:
  kustomize build /tmp/chartify1788651568/test --output /tmp/chartify1788651568/test/all.patched.yaml

OUTPUT:
  :: This is {Version:kustomize/v4.4.1 GitCommit:b2d65ddc98e09187a8e38adc27c30bab078c1dbf BuildDate:2021-11-11T23:36:27Z GoOs:linux GoArch:amd64}
  Error: accumulating resources: accumulation err='accumulating resources from 'helmx.1.rendered/chartName/templates/test.yaml': evalsymlink failure on '/tmp/chartify1788651568/test/helmx.1.rendered/chartName/templates/test.yaml' : lstat /tmp/chartify1788651568/test/helmx.1.rendered: no such file or directory': evalsymlink failure on '/tmp/chartify1788651568/test/helmx.1.rendered/chartName/templates/test.yaml' : lstat /tmp/chartify1788651568/test/helmx.1.rendered: no such file or directory]
```

**Comment by mumoshu** [comment]
@icy Hey! Thanks for reporting. It turned out to be a different one. I was able to reproduce it via a new integration test case and fixed it as you can see at https://github.com/variantdev/chartify/pull/29

**Comment by icy** [comment]
> @icy Hey! Thanks for reporting. It turned out to be a different one. I was able to reproduce it via a new integration test case and fixed it as you can see at #29

I confirm that works well now. Thanks a lot for your work.

**Comment by mumoshu** [comment]
@icy Thanks for your confirmation ☺️
